internal/detector/detector.go

```diff
@@ -77,23 +80,25 @@ func (m *Marker) DetectFunctionalUpstream(us *v1.Upstream) (*v1.ServiceInfo, map
 	// try every possible detector concurrently
 	for _, d := range m.detectors {
 		go func(d Interface) {
-			err := withBackoff(func() error {
-				serviceInfo, annotations, err := d.DetectFunctionalService(addr)
+			err := backoff.WithBackoff(func() error {
+				serviceInfo, annotations, err := d.DetectFunctionalService(us, addr)
 				if err != nil {
 					return err
 				}
 				// success
 				close(stop)
 				serviceInfoC <- serviceInfo
 				annotationsC <- annotations
+				m.m.Lock()
+				m.finishedOrFailed[us.Name] = maxRetries
```

Comment:
why is this set to maxRetries?

Reply:
if we reach this line we don't need to try the upstream anymore. this is my way of "toggling" the detector to *off* for this upstream

Reply:
I thought so but wasn't sure. So in this case the test about should be
already >=